controllers/upgrade.go

```diff
@@ -777,57 +840,167 @@ func (r *RollingUpgradeContext) ClusterBallooning(batchSize int) (bool, int) {
+				if err := r.uncordonAndRemoveAnnotation(node); err != nil {
+					r.Error(err, "failed to uncordon and remove annotation", "instanceID", instance.InstanceId, "nodeName", node.Name, "name", r.RollingUpgrade.NamespacedName())
+					errorCount++
+					// Continue with other nodes
+					continue
 				}
+				processedCount++
 			}
 		}
 	}
+
+	if cordonNode {
+		r.Info("early cordon operation completed", "processedNodes", processedCount, "errors", errorCount, "name", r.RollingUpgrade.NamespacedName())
+	} else {
+		r.Info("uncordon operation completed", "processedNodes", processedCount, "errors", errorCount, "name", r.RollingUpgrade.NamespacedName())
+	}
+
+	// Return error only if we had errors and processed no nodes successfully
+	if errorCount > 0 && processedCount == 0 {
+		return false, fmt.Errorf("failed to process any nodes: %d errors encountered", errorCount)
+	}
+
 	return true, nil
 }
+
+// cordonAndAnnotateNode cordons a node and adds annotation to track it was cordoned by upgrade-manager
+func (r *RollingUpgradeContext) cordonAndAnnotateNode(node *corev1.Node) error {
+	// First we cordon the node
+	if err := r.Auth.CordonUncordonNode(node, r.Auth.Kubernetes, true); err != nil {
+		return fmt.Errorf("failed to cordon node: %v", err)
+	}
+
+	// Retry annotation update with fresh node objects to handle resource version conflicts
+	maxRetries := 3
+	for attempt := 1; attempt <= maxRetries; attempt++ {
+		// Get fresh node object to avoid resource version conflicts
+		freshNode, err := r.Auth.Kubernetes.CoreV1().Nodes().Get(context.Background(), node.Name, metav1.GetOptions{})
+		if err != nil {
+			// If we can't get the fresh node, try to uncordon to avoid inconsistent state
+			if uncordonErr := r.Auth.CordonUncordonNode(node, r.Auth.Kubernetes, false); uncordonErr != nil {
+				r.Error(uncordonErr, "failed to rollback cordon after get failure", "nodeName", node.Name)
+			}
+			return fmt.Errorf("failed to get fresh node object: %v", err)
+		}
+
+		// Add annotation to track that we cordoned it
+		if freshNode.Annotations == nil {
+			freshNode.Annotations = make(map[string]string)
+		}
+		freshNode.Annotations[EarlyCordonAnnotationKey] = EarlyCordonAnnotationValue
+
+		// Update the node with the annotation
+		if _, err := r.Auth.Kubernetes.CoreV1().Nodes().Update(context.Background(), freshNode, metav1.UpdateOptions{}); err != nil {
+			if attempt < maxRetries && strings.Contains(err.Error(), "the object has been modified") {
+				r.Info("resource version conflict, retrying annotation", "nodeName", freshNode.Name, "attempt", attempt, "maxRetries", maxRetries)
+				time.Sleep(time.Duration(attempt) * 100 * time.Millisecond) // Exponential backoff
+				continue
+			}
+			// If annotation fails after all retries, try to uncordon to avoid leaving node in inconsistent state
+			if uncordonErr := r.Auth.CordonUncordonNode(freshNode, r.Auth.Kubernetes, false); uncordonErr != nil {
```

Comment:
Probably you can handle al the uncordon scenarios in uncordonAndRemoveAnnotation() function. Just add a a check in the uncordonAndRemoveAnnotation() function to ignore annotations if they are not present.

Reply:
That is a good suggestion, it would make uncordonAndRemoveAnnotation handle all uncordoning scenarios